### PR TITLE
Extract vector fusion, enhance book deduplication, standardize config

### PR DESCRIPTION
**Pull request by WilliamAGH**
## Summary

Strengthens search result deduplication and extracts the embedding-fusion vector math into a dedicated component.

## Search deduplication

- `CandidateKeyResolver` now keys candidates on a content-first priority chain: **ISBN-13 → ISBN-10 → normalized title+author → internal ID**, so persisted rows and external-fallback rows collapse into one candidate when providers assign different IDs to the same work.
- `SearchPageAssembler` dedupes on the resolved content key rather than `book.id`.
- `SearchPaginationService.deduplicateCandidates` uses the same resolver so persisted/fallback merging and streaming pages agree on identity.

## Regression avoidance — cover filtering before dedupe

Content-key dedupe can collapse a persisted book (e.g. low-quality cover) with its fallback-provider twin (better cover). With the previous filter-after-dedupe order, first-seen-wins could drop the better candidate and then filter out the survivor entirely. `SearchPageAssembler.buildPage` now applies cover/resolution filtering **before** the dedupe loop, so only eligible candidates enter the `LinkedHashMap` and the first survivor per key is guaranteed to serve the request.

## Key-resolver hardening

- Regex patterns in `normalizeText` lifted to `static final Pattern` constants to avoid per-call recompilation on the search hot path.
- Normalization retains `+` and `#` so programming-language titles like `C++`, `C#`, `F#` stay distinct.
- Partial-data rows (missing title or author with no ID) now return `Optional.empty()` instead of emitting asymmetric keys that could collapse unrelated works.
- Removed a dead `candidateKey.isEmpty()` guard in the assembler (the upstream `book.getId()` non-blank check guarantees `resolve()` always yields a key).

## Embedding vector fusion refactor

- New `BookSimilarityVectorFusion` `@Component` owns per-section L2-normalization and weighted fusion.
- `BookSimilarityEmbeddingService` delegates fusion and no longer carries vector math, tightening SRP and making fusion independently testable.

## Testing & tooling

- Added coverage for cross-provider title+author dedupe and unclustered Postgres duplicates that share title+author.
- Consolidated `.config/lefthook.yml` whitespace and end-of-file hooks into a single `whitespace-normalize` command to eliminate a race.

## Test plan

- [x] `./gradlew test` — green
- [x] Targeted tests for `service.SearchPagination*Test`, `support.search.*`, `application.similarity.*`, `domain.similarity.*` pass
- [ ] Manual smoke on `/search?q=...` for a query known to return persisted + fallback duplicates, verifying a single merged candidate per work and no missing results under cover-preference filters

<h3>Greptile Summary</h3>

This PR strengthens book search deduplication by refactoring `CandidateKeyResolver` to prefer a normalized title+author key over provider/internal IDs, and reorders cover-preference filtering to run before deduplication in `SearchPageAssembler` so that a lower-quality duplicate can't suppress a richer candidate. It also extracts vector fusion math from `BookSimilarityEmbeddingService` into a new `@Component` (`BookSimilarityVectorFusion`) with no change to the underlying algorithm, and consolidates lefthook pre-commit commands.

<h3>Confidence Score: 5/5</h3>

Safe to merge — only P2 style findings, no logic errors or regressions identified

All findings are P2 (style/best-practice): an accidentally widened constant visibility and a potential apostrophe normalization edge case. The core deduplication logic, vector fusion refactor, and test coverage are correct. No P0 or P1 issues found.

CandidateKeyResolver.java — apostrophe normalization trade-off worth a deliberate decision; BookSimilarityVectorFusion.java — EMBEDDING_DIMENSION visibility

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/main/java/net/findmybook/support/search/CandidateKeyResolver.java | Priority chain reordered to prefer title+author over ID; Pattern constants pre-compiled correctly; apostrophe normalization may cause cross-provider dedup misses (P2) |
| src/main/java/net/findmybook/support/search/SearchPageAssembler.java | Cover filtering reordered to run before deduplication; dedup now uses resolved content keys; previously-noted unreachable guard remains but is pre-existing |
| src/main/java/net/findmybook/application/similarity/BookSimilarityVectorFusion.java | Clean extraction of vector fusion math; EMBEDDING_DIMENSION accidentally widened to package-private (was private); logic is otherwise identical to the removed inlined methods |
| src/main/java/net/findmybook/application/similarity/BookSimilarityEmbeddingService.java | Removed EMBEDDING_DIMENSION constant and fusion/normalization methods; delegates to injected BookSimilarityVectorFusion; wiring change is straightforward |
| src/test/java/net/findmybook/service/SearchPaginationServicePagingTest.java | New test correctly verifies Postgres unclustered title+author deduplication with 'by ' prefix normalization |
| src/test/java/net/findmybook/service/SearchPaginationServiceFallbackTest.java | New fallback dedup test and verifyNoInteractions replacements look correct; assertions match the new resolver priority chain |
| .config/lefthook.yml | Trailing-whitespace and end-of-file checks consolidated into a single command; behavior is equivalent |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[rawResults] --> B{book != null\n& hasText id?}
    B -- No --> DROP1[dropped]
    B -- Yes --> C[applyCoverPreferences]
    C --> D{passes cover\n& resolution filter?}
    D -- No --> DROP2[dropped]
    D -- Yes --> E[CandidateKeyResolver.resolve]
    E --> F{ISBN-13?}
    F -- Yes --> K1[ISBN13: key]
    F -- No --> G{ISBN-10?}
    G -- Yes --> K2[ISBN10: key]
    G -- No --> H{normalizedTitle\n& normalizedAuthor\nboth non-blank?}
    H -- Yes --> K3[TITLE_AUTHOR: key]
    H -- No --> I{hasText id?}
    I -- Yes --> K4[ID: key]
    I -- No --> K5[Optional.empty - skip]
    K1 --> J{key already\nin map?}
    K2 --> J
    K3 --> J
    K4 --> J
    J -- Yes --> DROP3[duplicate dropped]
    J -- No --> L[add to orderedCandidatesByKey]
    L --> M[sort by primarySort + coverPriority]
    M --> N[slice to page window]
    N --> O[SearchPage returned]
```

<a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Asrc%2Fmain%2Fjava%2Fnet%2Ffindmybook%2Fapplication%2Fsimilarity%2FBookSimilarityVectorFusion.java%3A22%0A**%60EMBEDDING_DIMENSION%60%20visibility%20widened%20unintentionally**%0A%0AThe%20constant%20was%20%60private%20static%20final%20int%60%20in%20the%20old%20%60BookSimilarityEmbeddingService%60%20and%20is%20now%20package-private%20%28no%20modifier%29%20in%20%60BookSimilarityVectorFusion%60.%20It's%20only%20referenced%20inside%20the%20class%2C%20so%20the%20broader%20visibility%20is%20an%20accidental%20byproduct%20of%20the%20extraction.%20Either%20make%20it%20%60private%20static%20final%60%20to%20match%20its%20actual%20usage%2C%20or%20%60public%20static%20final%60%20if%20external%20callers%20%28tests%2C%20validators%29%20are%20expected%20to%20reference%20it.%0A%0A%60%60%60suggestion%0A%20%20%20%20private%20static%20final%20int%20EMBEDDING_DIMENSION%20%3D%202560%3B%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Asrc%2Fmain%2Fjava%2Fnet%2Ffindmybook%2Fsupport%2Fsearch%2FCandidateKeyResolver.java%3A31%0A**Apostrophes%20stripped%20to%20spaces%2C%20causing%20asymmetric%20normalization%20across%20providers**%0A%0A%60NON_SEARCH_TEXT_PATTERN%60%20%28%60%5B%5E%5Cp%7BL%7D0-9%2B%23%5D%2B%60%29%20replaces%20apostrophes%20with%20a%20space%2C%20so%20%22O'Brien%22%20normalizes%20to%20%22o%20brien%22%20while%20a%20provider%20that%20has%20already%20stripped%20apostrophes%20would%20give%20%22OBrien%22%20%E2%86%92%20%22obrien%22.%20These%20produce%20different%20keys%2C%20silently%20preventing%20deduplication%20for%20the%20same%20author%20when%20providers%20differ%20in%20apostrophe%20handling.%20The%20same%20applies%20to%20book%20titles%20%28e.g.%2C%20%22Shakespeare's%20Plays%22%20vs%20%22Shakespeares%20Plays%22%29.%0A%0AIf%20the%20intent%20is%20that%20only%20exact-format%20matches%20deduplicate%2C%20document%20this%20limitation%20so%20callers%20understand%20that%20cross-provider%20apostrophe%20differences%20are%20out%20of%20scope.%0A%0A&repo=williamagh%2Ffindmybook&pr=92&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=2"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=2">Fix All in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/cursor?prompt=Fix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Asrc%2Fmain%2Fjava%2Fnet%2Ffindmybook%2Fapplication%2Fsimilarity%2FBookSimilarityVectorFusion.java%3A22%0A**%60EMBEDDING_DIMENSION%60%20visibility%20widened%20unintentionally**%0A%0AThe%20constant%20was%20%60private%20static%20final%20int%60%20in%20the%20old%20%60BookSimilarityEmbeddingService%60%20and%20is%20now%20package-private%20%28no%20modifier%29%20in%20%60BookSimilarityVectorFusion%60.%20It's%20only%20referenced%20inside%20the%20class%2C%20so%20the%20broader%20visibility%20is%20an%20accidental%20byproduct%20of%20the%20extraction.%20Either%20make%20it%20%60private%20static%20final%60%20to%20match%20its%20actual%20usage%2C%20or%20%60public%20static%20final%60%20if%20external%20callers%20%28tests%2C%20validators%29%20are%20expected%20to%20reference%20it.%0A%0A%60%60%60suggestion%0A%20%20%20%20private%20static%20final%20int%20EMBEDDING_DIMENSION%20%3D%202560%3B%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Asrc%2Fmain%2Fjava%2Fnet%2Ffindmybook%2Fsupport%2Fsearch%2FCandidateKeyResolver.java%3A31%0A**Apostrophes%20stripped%20to%20spaces%2C%20causing%20asymmetric%20normalization%20across%20providers**%0A%0A%60NON_SEARCH_TEXT_PATTERN%60%20%28%60%5B%5E%5Cp%7BL%7D0-9%2B%23%5D%2B%60%29%20replaces%20apostrophes%20with%20a%20space%2C%20so%20%22O'Brien%22%20normalizes%20to%20%22o%20brien%22%20while%20a%20provider%20that%20has%20already%20stripped%20apostrophes%20would%20give%20%22OBrien%22%20%E2%86%92%20%22obrien%22.%20These%20produce%20different%20keys%2C%20silently%20preventing%20deduplication%20for%20the%20same%20author%20when%20providers%20differ%20in%20apostrophe%20handling.%20The%20same%20applies%20to%20book%20titles%20%28e.g.%2C%20%22Shakespeare's%20Plays%22%20vs%20%22Shakespeares%20Plays%22%29.%0A%0AIf%20the%20intent%20is%20that%20only%20exact-format%20matches%20deduplicate%2C%20document%20this%20limitation%20so%20callers%20understand%20that%20cross-provider%20apostrophe%20differences%20are%20out%20of%20scope.%0A%0A&pr=92&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursorDark.svg?v=2"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursor.svg?v=2">Fix All in Cursor</picture></a> <a href="https://chatgpt.com/codex/deeplink?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22williamagh%2Ffindmybook%22%20on%20the%20existing%20branch%20%22dev%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22dev%22.%0A%0AFix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Asrc%2Fmain%2Fjava%2Fnet%2Ffindmybook%2Fapplication%2Fsimilarity%2FBookSimilarityVectorFusion.java%3A22%0A**%60EMBEDDING_DIMENSION%60%20visibility%20widened%20unintentionally**%0A%0AThe%20constant%20was%20%60private%20static%20final%20int%60%20in%20the%20old%20%60BookSimilarityEmbeddingService%60%20and%20is%20now%20package-private%20%28no%20modifier%29%20in%20%60BookSimilarityVectorFusion%60.%20It's%20only%20referenced%20inside%20the%20class%2C%20so%20the%20broader%20visibility%20is%20an%20accidental%20byproduct%20of%20the%20extraction.%20Either%20make%20it%20%60private%20static%20final%60%20to%20match%20its%20actual%20usage%2C%20or%20%60public%20static%20final%60%20if%20external%20callers%20%28tests%2C%20validators%29%20are%20expected%20to%20reference%20it.%0A%0A%60%60%60suggestion%0A%20%20%20%20private%20static%20final%20int%20EMBEDDING_DIMENSION%20%3D%202560%3B%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Asrc%2Fmain%2Fjava%2Fnet%2Ffindmybook%2Fsupport%2Fsearch%2FCandidateKeyResolver.java%3A31%0A**Apostrophes%20stripped%20to%20spaces%2C%20causing%20asymmetric%20normalization%20across%20providers**%0A%0A%60NON_SEARCH_TEXT_PATTERN%60%20%28%60%5B%5E%5Cp%7BL%7D0-9%2B%23%5D%2B%60%29%20replaces%20apostrophes%20with%20a%20space%2C%20so%20%22O'Brien%22%20normalizes%20to%20%22o%20brien%22%20while%20a%20provider%20that%20has%20already%20stripped%20apostrophes%20would%20give%20%22OBrien%22%20%E2%86%92%20%22obrien%22.%20These%20produce%20different%20keys%2C%20silently%20preventing%20deduplication%20for%20the%20same%20author%20when%20providers%20differ%20in%20apostrophe%20handling.%20The%20same%20applies%20to%20book%20titles%20%28e.g.%2C%20%22Shakespeare's%20Plays%22%20vs%20%22Shakespeares%20Plays%22%29.%0A%0AIf%20the%20intent%20is%20that%20only%20exact-format%20matches%20deduplicate%2C%20document%20this%20limitation%20so%20callers%20understand%20that%20cross-provider%20apostrophe%20differences%20are%20out%20of%20scope.%0A%0A"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=2"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=2">Fix All in Codex</picture></a>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: src/main/java/net/findmybook/application/similarity/BookSimilarityVectorFusion.java
Line: 22

Comment:
**`EMBEDDING_DIMENSION` visibility widened unintentionally**

The constant was `private static final int` in the old `BookSimilarityEmbeddingService` and is now package-private (no modifier) in `BookSimilarityVectorFusion`. It's only referenced inside the class, so the broader visibility is an accidental byproduct of the extraction. Either make it `private static final` to match its actual usage, or `public static final` if external callers (tests, validators) are expected to reference it.

```suggestion
    private static final int EMBEDDING_DIMENSION = 2560;
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: src/main/java/net/findmybook/support/search/CandidateKeyResolver.java
Line: 31

Comment:
**Apostrophes stripped to spaces, causing asymmetric normalization across providers**

`NON_SEARCH_TEXT_PATTERN` (`[^\p{L}0-9+#]+`) replaces apostrophes with a space, so "O'Brien" normalizes to "o brien" while a provider that has already stripped apostrophes would give "OBrien" → "obrien". These produce different keys, silently preventing deduplication for the same author when providers differ in apostrophe handling. The same applies to book titles (e.g., "Shakespeare's Plays" vs "Shakespeares Plays").

If the intent is that only exact-format matches deduplicate, document this limitation so callers understand that cross-provider apostrophe differences are out of scope.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (2): Last reviewed commit: ["fix(search): filter cover preferences be..."](https://github.com/williamagh/findmybook/commit/932853b33de934659b530ee36cd878ee781b9ef9) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=29653238)</sub>

### CODE REVIEW

File: src/main/java/net/findmybook/support/search/CandidateKeyResolver.java [hunk 1]
```diff
@@ -10,12 +10,21 @@
...
  */
 public final class CandidateKeyResolver {
 
+    private static final String ISBN13_KEY_PREFIX = "ISBN13:";
+    private static final String ISBN10_KEY_PREFIX = "ISBN10:";
+    private static final String TITLE_AUTHOR_KEY_PREFIX = "TITLE_AUTHOR:";
+    private static final String ID_KEY_PREFIX = "ID:";
+    private static final String CONTRIBUTOR_BY_PREFIX = "by ";
+    private static final String NON_SEARCH_TEXT_PATTERN = "[^\\p{L}0-9]+";
+    private static final String WHITESPACE_PATTERN = "\\s+";
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/support/search/CandidateKeyResolver.java]: # Suggestion Implemented ✅

Accepted. Converted `NON_SEARCH_TEXT_PATTERN` and `WHITESPACE_PATTERN` to `static final Pattern` constants and switched `normalizeText` to `matcher().replaceAll(" ")`. Imported `java.util.regex.Pattern`.

Validation: `./gradlew test` — BUILD SUCCESSFUL.

Commit: 7442bdf595c815f61b5546f2ed10a589b2f20ba8

File: src/main/java/net/findmybook/support/search/SearchPageAssembler.java [hunk 2]
```diff
@@ -52,19 +52,27 @@ public SearchPaginationService.SearchPage buildPage(String query,
...
         for (Book book : safeRawResults) {
-            if (book == null || !StringUtils.hasText(book.getId()) || ordered.containsKey(book.getId())) {
+            if (book == null || !StringUtils.hasText(book.getId())) {
                 continue;
             }
-            ordered.put(book.getId(), book);
+            Optional<String> candidateKey = CandidateKeyResolver.resolve(book);
+            if (candidateKey.isEmpty()) {
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/support/search/SearchPageAssembler.java]: # Suggestion Implemented ✅

Accepted per [CLN1b]. Removed the `candidateKey.isEmpty()` branch: after filtering `book.getId()` to non-blank up the stack, `resolve(book)` always surfaces a key (ISBN→title+author→ID path).

Also reorganized so cover-preference filtering runs **before** dedupe (see sibling threads on the same file) — the dedupe loop now only sees candidates that will actually survive filtering.

Commit: 932853b33de934659b530ee36cd878ee781b9ef9

File: src/main/java/net/findmybook/application/similarity/BookSimilarityVectorFusion.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            if (embedding == null) {
+                throw new IllegalStateException("Missing section embedding for " + sectionInput.sectionKey().key());
+            }
+            double norm = vectorNorm(embedding);
+            if (norm == 0.0d) {
+                continue;
+            }
+            double weight = weights.get(sectionInput.sectionKey());
+            for (int index = 0; index < EMBEDDING_DIMENSION; index++) {
+                fused[index] += (embedding.get(index) / norm) * weight;
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/application/similarity/BookSimilarityVectorFusion.java:52]: # Suggestion Analysis 🔍

Rejected — the NPE cannot occur given the current invariants upstream.

**Source evidence:** `BookSimilaritySourceDocument.create` ([BookSimilaritySourceDocument.java:78-93](https://github.com/WilliamAGH/findmybook/blob/dev/src/main/java/net/findmybook/domain/similarity/BookSimilaritySourceDocument.java#L78-L93)) filters `sectionInputs` with the same gate the fusion policy applies:

```java
if (activeProfile.weightFor(sectionKey) <= 0.0d) {
    continue;
}
```

So every `sectionInput` in `sourceDocument.sectionInputs()` has a positive active-profile weight. `BookSimilarityFusionPolicy.normalizedWeightsFor` ([BookSimilarityFusionPolicy.java:85-111](https://github.com/WilliamAGH/findmybook/blob/dev/src/main/java/net/findmybook/domain/similarity/BookSimilarityFusionPolicy.java#L85-L111)) applies the same `weight > 0.0d` filter, so for any `sectionKey` present in both `sectionEmbeddings.keySet()` (the input) and `sourceDocument.sectionInputs()`, the returned map contains that key.

The only residual divergence — a `sectionInput.sectionKey()` missing from `sectionEmbeddings` — is already caught by the explicit `if (embedding == null)` guard at [BookSimilarityVectorFusion.java:43](https://github.com/WilliamAGH/findmybook/blob/dev/src/main/java/net/findmybook/application/similarity/BookSimilarityVectorFusion.java#L43), which throws `IllegalStateException` with context.

Adding an extra null-guard on `weights.get(...)` would introduce a silent `continue` masking an upstream invariant break — contrary to [BLK1a]/[BLK1c] (no silent fallbacks, fix root causes).

File: src/main/java/net/findmybook/support/search/CandidateKeyResolver.java [hunk 4]
```diff
@@ -31,28 +40,55 @@ public static Optional<String> resolve(Book book) {
...
 
-        String title = Optional.ofNullable(book.getTitle()).orElse("").trim().toLowerCase(Locale.ROOT);
-        String firstAuthor = (book.getAuthors() == null || book.getAuthors().isEmpty())
-            ? ""
-            : Optional.ofNullable(book.getAuthors().get(0)).orElse("").trim().toLowerCase(Locale.ROOT);
-
         if (!title.isBlank() || !firstAuthor.isBlank()) {
-            return Optional.of("TITLE_AUTHOR:" + title + "::" + firstAuthor);
+            return Optional.of(TITLE_AUTHOR_KEY_PREFIX + title + "::" + firstAuthor);
         }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/support/search/CandidateKeyResolver.java]: # Suggestion Implemented ✅

Accepted. The partial-data branch now returns `Optional.empty()` instead of emitting asymmetric `TITLE_AUTHOR:foo::` / `TITLE_AUTHOR:::bar` keys that would collapse distinct works.

Also updated the `resolve()` Javadoc to document this contract so upstream callers know partial-data rows are dropped intentionally.

Commit: 7442bdf595c815f61b5546f2ed10a589b2f20ba8

File: src/main/java/net/findmybook/support/search/SearchPageAssembler.java [hunk 5]
```diff
@@ -52,19 +52,27 @@ public SearchPaginationService.SearchPage buildPage(String query,
...
+            orderedCandidatesByKey.put(resolvedKey, book);
             insertionOrder.put(book.getId(), position++);
         }
 
-        List<Book> uniqueResults = new ArrayList<>(ordered.values());
+        List<Book> uniqueResults = new ArrayList<>(orderedCandidatesByKey.values());
         List<Book> filtered = applyCoverPreferences(uniqueResults, coverSource, resolutionPreference);
         applyAuthorIntentPenalties(filtered, query);
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/support/search/SearchPageAssembler.java]: # Suggestion Implemented ✅

Accepted. `SearchPageAssembler.buildPage` now applies `applyCoverPreferences` **before** the dedupe loop. The dedupe `LinkedHashMap` only sees candidates that will actually survive filtering, so a better-covered fallback twin can no longer be silently dropped by a lower-quality persisted duplicate.

Validation: `./gradlew test` — BUILD SUCCESSFUL.

Commit: 932853b33de934659b530ee36cd878ee781b9ef9

File: src/main/java/net/findmybook/support/search/CandidateKeyResolver.java [hunk 6]
```diff
@@ -31,28 +40,55 @@ public static Optional<String> resolve(Book book) {
...
+
+    private static String normalizeText(String rawText) {
+        if (!StringUtils.hasText(rawText)) {
+            return "";
+        }
+        return rawText
+            .toLowerCase(Locale.ROOT)
+            .replaceAll(NON_SEARCH_TEXT_PATTERN, " ")
+            .trim()
+            .replaceAll(WHITESPACE_PATTERN, " ");
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/support/search/CandidateKeyResolver.java]: # Suggestion Implemented ✅

Accepted. `NON_SEARCH_TEXT_PATTERN` and `WHITESPACE_PATTERN` are now `static final Pattern` constants, and `normalizeText` uses `matcher().replaceAll(" ")` to avoid per-call regex recompilation on search hot paths.

Commit: 7442bdf595c815f61b5546f2ed10a589b2f20ba8

File: src/main/java/net/findmybook/support/search/SearchPageAssembler.java [hunk 7]
```diff
@@ -52,19 +52,27 @@ public SearchPaginationService.SearchPage buildPage(String query,
...
                 continue;
             }
-            ordered.put(book.getId(), book);
+            Optional<String> candidateKey = CandidateKeyResolver.resolve(book);
+            if (candidateKey.isEmpty()) {
+                continue;
+            }
+            String resolvedKey = candidateKey.get();
+            if (orderedCandidatesByKey.containsKey(resolvedKey)) {
+                continue;
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/support/search/SearchPageAssembler.java]: # Suggestion Implemented ✅

Accepted — valid P1. This PR introduced the issue: before the normalized title+author key, dedupe was by `book.id`, so persisted and fallback rows kept separate entries and each survived or failed cover filtering on its own. After collapsing them on content keys, first-seen-wins semantics can suppress the better-covered candidate, which then disappears when the remaining lower-quality row fails cover filtering.

Fix: `SearchPageAssembler.buildPage` now applies `applyCoverPreferences` to candidates **before** the dedupe loop. Only cover-eligible candidates enter the `LinkedHashMap`, so the first surviving candidate per key is guaranteed to serve the request.

Commit: 932853b33de934659b530ee36cd878ee781b9ef9

File: src/main/java/net/findmybook/support/search/CandidateKeyResolver.java [hunk 8]
```diff
@@ -31,28 +40,55 @@ public static Optional<String> resolve(Book book) {
...
+        return normalized;
+    }
+
+    private static String normalizeText(String rawText) {
+        if (!StringUtils.hasText(rawText)) {
+            return "";
+        }
+        return rawText
+            .toLowerCase(Locale.ROOT)
+            .replaceAll(NON_SEARCH_TEXT_PATTERN, " ")
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/support/search/CandidateKeyResolver.java]: # Suggestion Implemented ✅

Accepted. Extended the retained character class in `NON_SEARCH_TEXT_PATTERN` from `[^\p{L}0-9]+` to `[^\p{L}0-9+#]+` so `+` and `#` survive normalization. Programming-language titles like `C++`, `C#`, `F#` now produce distinct dedupe keys (`c++`, `c#`, `f#`) rather than collapsing to `c`/`f`.

Commit: 7442bdf595c815f61b5546f2ed10a589b2f20ba8
